### DISCUSSION TITLE
Mini game for ELKS Basic

### DISCUSSION THREAD

**Discussion by tyama501** [jbruchon/elks#1618] [Show and tell]
Hello,

I wrote a mini game for ELKS basic example.
It uses graphic functions, so it is only for PC-98 for now.

https://github.com/tyama501/microef

For making game,
I think it is better to have the following functions for the Basic.
(1) "ELSE" for "IF"
There is no "ELSE", so it is a little bit hard to code.
(2) RANDOMISE
RANDOMISE is not implemented in the basic yet,
so always RND is same sequence when starting the basic.
(3) File access.
I don't know what functions are available in Sinclair basic,
but writing/reading from files are needed to save/load.

Thank you.

**Comment by tyama501** [comment]
Here is the playing movie.

https://github.com/jbruchon/elks/assets/61556504/a6af427e-acfc-4b14-816e-9955b785f1f7

**Comment by ghaerr** [comment]
Hello @tyama501,

Your new game looks pretty neat!

It looks like it might be nice to have a function to turn off/on the cursor, or perhaps that could be automatic when graphics is initialized.

You're welcome to add `RANDOMIZE n` function using C `srand(n)`, where `n` is an optional parameter (otherwise use `time(NULL)`).

Looking through various Sinclair BASIC docs, I'm not finding any implementation of ELSE. Its implementation might be a bit complicated given that multiple "statements" using `:` can be used after THEN.

Sinclair doesn't support disk I/O, but various other BASICs use `OPEN #1,"file"` then `PRINT #1,...` etc (or something similar). We could look into adding that if you would like.

Thank you!

**Comment by ghaerr** [comment]
I just found that Sinclair BASIC does support SAVE/LOAD for DATA arrays of integer and string, so this might be easier to implement than previous stated:

![save](https://github.com/jbruchon/elks/assets/11985637/978700d3-2c73-4942-9a61-fe76fa3c35cd)

**Comment by tyama501** [comment]
Thank you @ghaerr ,

Yes, I think it is useful to have a command to enable/disable cursors from basic.
Well, I forgot how we are handling cursors when direct console.
There is a conio-pc98-asm.S and there are cursors on/off using Biso int 0x18 for pc-98 but 
I think this is only used in Headless console in conio-pc98.c.
Is the elks drawing cursors when  when direct console?

I will try to add RANDMIZE.

**Comment by ghaerr** [comment]
ELKS uses the underlying hardware to draw the cursor, which is currently always left default ON using the power on cursor shape/size.

We should probably add the ANSI cursor on/off sequence to IBM PC direct console, and it can then program the 6845 CRT controller to enable or disable the hardware cursor, and use INT 18h for PC-98 direct console?

ANSI `ESC [ ? 25 l` hides the cursor and `ESC [ ? 25 h` shows the cursor.

If you implement the PC-98 cursor on/off, I can add IBM PC support via 6845 chip.

**Reply by tyama501** [reply]
Oh, there are some functions like PositionCursor and others in the direct console, that is why I thought elks somehow writing cursor to the console.

Hmm, then there should be enabling pc-98 cursor somewhere. I will look again later.

**Reply by tyama501** [reply]
I see. kbd_init is calling conio_init, and there is cursor on.

**Reply by tyama501** [reply]
Is the function AnsiCmd in console.c that I need to add the ansi sequence?
Also adding function for on/off cursor in console-direct-pc98 and calling from the ansi sequence?

**Reply by ghaerr** [reply]
Yes, AnsiCmd will work. That function should continue adding characters to C->params[] until `h` or `l` seen, after which you'll have to check for first character being `?` then parameter value = 25 for call to `cursor_on/cursor_off`. You may have to add empty functions for IBM PC for the cursor and I can add 6845 CRTC code after PC-98 is working.

**Reply by tyama501** [reply]
Hello @ghaerr ,

If I add code something like this,
it seems that the console exits the sequence before pressing the number key after pressing 'ESC [ ?' keys.
Am I doing something wrong?
I tried to add printk for debug but if I add that the console did not work properly.

    case 'h':                   /* cursor on */
      {
        char *p = (char *)C->params;
        if (*p == '?')
            p++;
        else
            break;
        n = atoi(p);
        if (n == 25)
            CursorOn(C);
      }
      break;
    case 'l':                   /* cursor off */
      {
        char *p = (char *)C->params;
        if (*p == '?')
            p++;
        else
            break;
        n = atoi(p);
        if (n == 25)
            CursorOff(C);
      }
      break;

Thank you.

**Reply by ghaerr** [reply]
Hello @tyama501,

I cannot see anything coded incorrectly with your example. It appears if `isalpha('?')` were to return false, then the console would exit the sequence before a terminating alpha character is seen. I just tested `isalpha` (outside of ELKS on macOS using the same `#define` definition) and it appears to work correctly. Perhaps it is not working on ELKS, we would need to inspect the code generated or add some debug statements.

In a final version, it might be better to consolidate `CursorOn` and `CursorOff` into a single `DisplayCursor(int onoff)`, which would also allow for the `m` and `l` cases to be combined for less code duplication.

For debugging, you might be able to configure the build for serial console, which would allow printk to serial port, while the PC-98 console was used for testing the ANSI sequences.

Thank you!

**Reply by tyama501** [reply]
Hello @ghaerr ,

I have one question about the following codes.
In esc_char(), it sets C->fsm to "AnsiCmd" when case '[' but
it gets back to C->fsm = std_char if c is not '[' or '7' or '8'.
Is this ok to accumulate remaining characters?

static void esc_char(register Console * C, int c)
{
    /* Parse CSI sequence */
    C->parmptr = C->params;
    switch (c) {
    case '[':
        C->fsm = AnsiCmd;
        return;
    case '7':           /* DEC save cursor pos */
        C->savex = C->cx;
        C->savey = C->cy;
        break;
    case '8':           /* DEC restore cursor pos */
        C->cx = C->savex;
        C->cy = C->savey;
        break;
    }
    C->fsm = std_char;
}

**Reply by ghaerr** [reply]
Hello @tyama501, I think that CSI parsing is OK, since that code handles `ESC 7` and `ESC 8` only.  

It seems we need a better way to debug console since its hard to insert printk in the switch statements. Let me take a quick look at what might be happening using the original code you posted above.

**Reply by ghaerr** [reply]
Hello @tyama501,

I just tested your original proposed code in `AnsiCmd`, and it works for both `ESC [ ? 25 h` and `ESC [ ? 25 l`. Thus, I think the problem is that your functions `CursorOn` and `CursorOff` are faulty.

I was able to test your code by adding it to console.c and replace the call to CursorOn/CursorOff with `C->attr = A_REVERSE`. Then, logging in to the `sash` shell (using `toor`) which doesn't use linenoise thus not affecting ESC input/output, I typed in `ESC [ ? 25 h` at the shell prompt and the output turned reverse video, showing success. (Remember to `rm *.o` in drivers/char to force console.c to be recompiled since it is included into another .c file).

After fixing your cursor on/off routines, I still recommend possibly combining the functions in AnsiCmd and also using `DisplayCursor(int onff)` for a final version.

Thank you!

**Reply by tyama501** [reply]
Hello @ghaerr 

Thank you.

I tried sash.
It woked with sash...

Do I need to do something to work with ash?

**Reply by ghaerr** [reply]
> Do I need to do something to work with ash?

We're only talking about the ability to for a person to input escape sequences interactively to a shell, not the ability for the escape sequences to be interpreted by the kernel (which always occurs if CONFIG_EMUL_ANSI is set and direct console is used). `ash` uses the `linenoise` library which interprets input escape sequences (such as arrow keys), so it intercepts sequences instead of passing them on to the kernel. Thus, `ash` cannot be used for testing. `sash` does not interpret any typed input for itself (other than full command lines), thus all the bytes gets passed to the kernel.

**Reply by tyama501** [reply]
Oh, ok. So this is correct behavior.

I will clean codes to use single DisplayCursor and
make PR.

It seems that I can run sash after booting with ash by typing "sash" so it would be ok.

Thank you.
